cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -204,6 +206,73 @@ public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.Workflo
     return programToRunRecord.values();
   }
 
+  @Nullable
+  WorkflowRunRecord getRecord(Id.Workflow id, String pid) {
+    RunId runId = RunIds.fromString(pid);
+    long startTime = RunIds.getTime(runId, TimeUnit.SECONDS);
+    MDSKey mdsKey = new MDSKey.Builder().add(id.getNamespaceId())
+      .add(id.getApplicationId()).add(id.getId()).add(startTime).build();
+    byte[] startRowKey = mdsKey.getKey();
+
+    Row indexRow = table.get(startRowKey);
+    if (indexRow == null) {
+      return null;
+    }
+    Map<byte[], byte[]> columns = indexRow.getColumns();
+    String workflowRunId = Bytes.toString(columns.get(RUNID));
+    long timeTaken = Bytes.toLong(columns.get(TIME_TAKEN));
+
+    List<ProgramRun> actionRunsList = GSON.fromJson(Bytes.toString(columns.get(NODES)), PROGRAM_RUNS_TYPE);
+    return new WorkflowRunRecord(workflowRunId, timeTaken, actionRunsList);
+  }
+
+  Map<String, WorkflowRunRecord> getDetailsOfRange(Id.Workflow workflow, String runId, int count, long timeInterval) {
+    Map<String, WorkflowRunRecord> mainRunRecords = getNeighbors(workflow, RunIds.fromString(runId),
+                                                                 count, timeInterval);
+    WorkflowRunRecord workflowRunRecord = getRecord(workflow, runId);
+    if (workflowRunRecord != null) {
+      mainRunRecords.put(workflowRunRecord.getWorkflowRunId(), workflowRunRecord);
+    }
+    return mainRunRecords;
+  }
+
+  private Map<String, WorkflowRunRecord> getNeighbors(Id.Workflow id, RunId runId, int count, long timeInterval) {
+    long startTime = RunIds.getTime(runId, TimeUnit.SECONDS);
+    Map<String, WorkflowRunRecord> workflowRunRecords = new HashMap<>();
+    int i = -count;
+    long prevStartTime = startTime - (timeInterval * count);
+    while (prevStartTime <= startTime + (count * timeInterval)) {
+      MDSKey mdsKey = new MDSKey.Builder().add(id.getNamespaceId())
+        .add(id.getApplicationId()).add(id.getId()).add(prevStartTime).build();
+      byte[] startRowKey = mdsKey.getKey();
+      Scan scan = new Scan(startRowKey, null);
+      Scanner scanner = table.scan(scan);
+      Row indexRow = scanner.next();
+      if (indexRow == null) {
+        return workflowRunRecords;
+      }
+      mdsKey = new MDSKey(indexRow.getRow());
+      MDSKey.Splitter splitter = mdsKey.split();
+      splitter.skipString();
+      splitter.skipString();
+      splitter.skipString();
+      long time = splitter.getLong();
+      if ((time >= (startTime - (count * timeInterval))) && time <= (startTime + (count * timeInterval))) {
```

Comment:
can you invert this condition?

```
if (!((time >= (startTime - (count * timeInterval))) && time <= (startTime + (count * timeInterval)))) {
  break;
}
```

This is to take out the else with a single break statement? Will reduce one level of indentation.